independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/CodestartCatalogLoader.java

```diff
@@ -6,57 +6,47 @@
 import com.fasterxml.jackson.databind.MapperFeature;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import io.quarkus.devtools.codestarts.core.CodestartSpec;
+import io.quarkus.devtools.codestarts.core.DefaultCodestartCatalog;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
 import java.util.Objects;
 import java.util.Set;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
-import org.apache.commons.io.FilenameUtils;
 
-final class CodestartLoader {
+public final class CodestartCatalogLoader {
+
     private static final ObjectMapper YAML_MAPPER = new ObjectMapper(new YAMLFactory())
             .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
             .enable(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY)
             .enable(DeserializationFeature.READ_ENUMS_USING_TO_STRING)
             .enable(MapperFeature.ACCEPT_CASE_INSENSITIVE_ENUMS);
 
-    private static final String CODESTARTS_DIR_BUNDLED = "bundled-codestarts";
-    private static final String CODESTARTS_DIR_FROM_EXTENSIONS = "codestarts";
-
-    private CodestartLoader() {
-    }
-
-    public static List<Codestart> loadAllCodestarts(CodestartInput input) throws IOException {
-        return loadAllCodestarts(input.getResourceLoader());
-    }
-
-    public static List<Codestart> loadAllCodestarts(CodestartResourceLoader resourceLoader) throws IOException {
-        return Stream.concat(loadBundledCodestarts(resourceLoader).stream(),
-                loadCodestartsFromExtensions(resourceLoader).stream()).collect(Collectors.toList());
-    }
-
-    public static Collection<Codestart> loadBundledCodestarts(CodestartInput input) throws IOException {
-        return loadBundledCodestarts(input.getResourceLoader());
-    }
-
-    public static Collection<Codestart> loadBundledCodestarts(CodestartResourceLoader resourceLoader) throws IOException {
-        return loadCodestarts(resourceLoader, CODESTARTS_DIR_BUNDLED);
+    private CodestartCatalogLoader() {
     }
 
-    public static Collection<Codestart> loadCodestartsFromExtensions(CodestartInput input)
+    public static DefaultCodestartCatalog loadDefaultCatalog(CodestartResourceLoader resourceLoader, String first,
```

Comment:
Return `CodestartCatalog`? Or is there something particular in this implementation that I am missing?

Reply:
@gastaldi I just wanted to avoid having the generic type everywhere. It really sucks that in Java you can't override a method parameter type with a subtype 😓

Reply:
@gastaldi any better idea?

Reply:
@gastaldi ok I removed `DefaultCodestartCatalog` and used `CodestartCatalog<CodestartProjectInput>`

Reply:
The solution is simple: don't use generics 😁. But that's OK

Reply:
@gastaldi how?